libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/CloudAdapter.cs

```diff
@@ -53,54 +58,219 @@ public CloudAdapter()
         {
-            if (httpRequest == null)
+            _ = httpRequest ?? throw new ArgumentNullException(nameof(httpRequest));
+            _ = httpResponse ?? throw new ArgumentNullException(nameof(httpResponse));
+            _ = bot ?? throw new ArgumentNullException(nameof(bot));
+
+            try
             {
-                throw new ArgumentNullException(nameof(httpRequest));
+                // Only GET requests for web socket connects are allowed
+                if (httpRequest.Method == HttpMethods.Get && httpRequest.HttpContext.WebSockets.IsWebSocketRequest)
+                {
+                    // All socket communication will be handled by the internal streaming-specific BotAdapter
+                    await ConnectAsync(httpRequest, bot, cancellationToken).ConfigureAwait(false);
+                }
+                else if (httpRequest.Method == HttpMethods.Post)
+                {
+                    // Deserialize the incoming Activity
+                    var activity = await HttpHelper.ReadRequestAsync<Activity>(httpRequest).ConfigureAwait(false);
+
+                    // A POST request must contain an Activity 
+                    if (string.IsNullOrEmpty(activity?.Type))
+                    {
+                        httpResponse.StatusCode = (int)HttpStatusCode.BadRequest;
+                        return;
+                    }
+
+                    // Grab the auth header from the inbound http request
+                    var authHeader = httpRequest.Headers["Authorization"];
+
+                    // Process the inbound activity with the bot
+                    var invokeResponse = await ProcessActivityAsync(authHeader, activity, bot.OnTurnAsync, cancellationToken).ConfigureAwait(false);
+
+                    // write the response, potentially serializing the InvokeResponse
+                    await HttpHelper.WriteResponseAsync(httpResponse, invokeResponse).ConfigureAwait(false);
+                }
+                else
+                {
+                    httpResponse.StatusCode = (int)HttpStatusCode.MethodNotAllowed;
+                }
+            }
+            catch (UnauthorizedAccessException)
+            {
+                // handle unauthorized here as this layer creates the http response
+                httpResponse.StatusCode = (int)HttpStatusCode.Unauthorized;
             }
+        }
 
-            if (httpResponse == null)
+        /// <summary>
+        /// Used to connect the adapter to a named pipe.
+        /// </summary>
+        /// <param name="pipeName">The name of the named pipe.</param>
+        /// <param name="bot">The bot instance to use.</param>
+        /// <param name="appId">The bot's application id.</param>
+        /// <param name="audience">The audience to use for outbound communication. This will vary by cloud environment.</param>
+        /// <param name="callerId">The callerId, this may be NULL.</param>
```

Comment:
You don't highlight in its `<param/>` that the `appId` can be null, though in the logic below it appears that this is acceptable.